api/src/main/java/com/github/eyefloaters/console/api/service/KafkaClusterService.java

```diff
@@ -105,33 +110,75 @@ public static Optional<Kafka> findCluster(SharedIndexInformer<Kafka> kafkaInform
                 .findFirst();
     }
 
-    public static Stream<ListenerStatus> externalListeners(Kafka kafka) {
-        return Optional.ofNullable(kafka.getStatus().getListeners())
-                .map(Collection::stream)
-                .orElseGet(Stream::empty)
-                .filter(l -> isExternalListener(kafka, l));
+    public static Optional<ListenerStatus> consoleListener(Kafka kafka) {
+        return kafka.getSpec().getKafka().getListeners().stream()
+            .filter(listener -> !KafkaListenerType.INTERNAL.equals(listener.getType()))
+            .filter(KafkaClusterService::supportedAuthentication)
+            .sorted((l1, l2) -> Integer.compare(listenerSortKey(l1), listenerSortKey(l2)))
+            .findFirst()
+            .map(listener -> listenerStatus(kafka, listener));
     }
 
-    public static boolean isExternalListener(Kafka kafka, ListenerStatus listener) {
-        return kafka.getSpec()
-                .getKafka()
-                .getListeners()
-                .stream()
-                .filter(sl -> sl.getName().equals(listener.getName()))
-                .map(GenericKafkaListener::getType)
-                .filter(Objects::nonNull)
-                .anyMatch(Predicate.not(KafkaListenerType.INTERNAL::equals));
+    static boolean supportedAuthentication(GenericKafkaListener listener) {
+        Optional<KafkaListenerAuthentication> listenerAuth = Optional.ofNullable(listener.getAuth());
+        String authType = listenerAuth.map(auth -> auth.getType()).orElse("");
+
+        if (authType.isBlank()) {
+            return true;
+        } else if (KafkaListenerAuthenticationOAuth.TYPE_OAUTH.equals(authType)) {
+            return true;
+        } else if (KafkaListenerAuthenticationCustom.TYPE_CUSTOM.equals(authType)) {
+            return listenerAuth
+                .map(KafkaListenerAuthenticationCustom.class::cast)
+                .filter(KafkaListenerAuthenticationCustom::isSasl)
+                .map(KafkaListenerAuthenticationCustom::getListenerConfig)
+                .map(listenerConfig -> listenerConfig.get("sasl.enabled.mechanisms"))
+                .map(mechanisms -> Arrays.asList(mechanisms.toString().toUpperCase(Locale.ROOT).split(",")))
+                .map(mechanisms -> mechanisms.contains("OAUTHBEARER"))
+                .orElse(false);
+        } else {
+            return false;
+        }
+    }
+
+    static int listenerSortKey(GenericKafkaListener listener) {
+        return annotatedListener(listener) ? -1 : 1;
+    }
+
+    static boolean annotatedListener(GenericKafkaListener listener) {
+        return Optional.ofNullable(listener.getConfiguration())
+            .map(GenericKafkaListenerConfiguration::getBootstrap)
+            .map(config -> config.getAnnotations())
+            .map(annotations -> annotations.get("eyefloaters.github.com/console-listener"))
```

Comment:
We should factor out constants for all annotations that we use. It helps with navigating the code later on. And we might not be eyefloaters for ever.

Reply:
Add an `Annotations` enum with (for now) this single value.